### ISSUE TITLE
dev-master does not use branch-alias when compared against conflict rules (bug?)

### ISSUE THREAD

**Issue by core23** [composer/composer#5505]
With the following composer.json:

``` json
{
    "name": "acme/test-bundle",
    "type": "symfony-bundle",
    "require": {
        "php": "^5.3 || ^7.0",
        "sonata-project/classification-bundle": "^4.0@dev",
        "sonata-project/media-bundle": "^4.0@dev"
    }
}
```

When I run this command:

```
composer update
```

I get this output:

```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for sonata-project/classification-bundle ^4.0@dev -> satisfiable by sonata-project/classification-bundle[4.x-dev].
    - don't install sonata-project/media-bundle 4.x-dev|install sonata-project/media-bundle dev-master
    - sonata-project/media-bundle dev-master conflicts with sonata-project/classification-bundle[4.x-dev].
    - Installation request for sonata-project/media-bundle ^4.0@dev -> satisfiable by sonata-project/media-bundle[4.x-dev].
```

This should resolve in installing both libraries in the latest `master` branch versions, because both have a branch-alias `"dev-master": "4.x-dev"`. Both library have a cyclic dependency ([media](https://github.com/sonata-project/SonataMediaBundle/blob/master/composer.json) and [classification](https://github.com/sonata-project/SonataClassificationBundle/blob/master/composer.json)).

The classification bundle has also a `conflict` setting, which prevents installing a version greater or lower 4.x. The upper rule seams to be the problem, because when I removed the upper conflict rule (`>= 5.0`) the dependencies could be resolved. Refs https://github.com/sonata-project/SonataClassificationBundle/pull/224.

**Comment by alcohol** [comment]
Cyclic dependencies are bad.

**Comment by alcohol** [comment]
It is not clear to me what is triggering this conflict rule though. Theoretically it should not. If I manually create the version and constraint, they do not satisfy each other. So, it looks like there might be a bug somewhere else in the code.

**Comment by alcohol** [comment]
I tried reproducing it with this fixture

```
--TEST--

See Github issue #5505 ( github.com/composer/composer/issues/5505 ).

--COMPOSER--
{
    "repositories": [
        {
            "type": "package",
            "package": [
                {
                    "name": "a",
                    "version": "4.x-dev",
                    "require-dev": { "b": "^3.0 || ^4.0@dev" },
                    "conflict": { "b": "<3.0 || >=5.0" }
                }, {
                    "name": "b",
                    "version": "4.x-dev",
                    "require": { "a": "^3.0 || ^4.0@dev" }
                }
            ]
        }
    ],
    "require": {
        "a": "^4.0@dev",
        "b": "^4.0@dev"
    }
}

--RUN--
install

--EXPECT--
Installing a (4.x-dev)
Installing b (4.x-dev)

--EXPECT-EXIT-CODE--
0
```

However, the installer correctly installs `4.x-dev` of both `a` and `b`.

**Comment by alcohol** [comment]
I think the problem is that dev-master internally is a version in the form of `9999999-dev`, which obviously conflicts with '<3.0 || >=5.0`.

I managed to reproduce it with the following fixture:

```
--TEST--

See Github issue #5505 ( github.com/composer/composer/issues/5505 ).

--COMPOSER--
{
    "repositories": [
        {
            "type": "package",
            "package": [
                {
                    "name": "a",
                    "version": "dev-master",
                    "require-dev": { "b": "^3.0 || ^4.0@dev" },
                    "conflict": { "b": "<3.0 || >=5.0" },
                    "extra": { "branch-alias": { "dev-master": "4.x-dev" } }
                }, {
                    "name": "b",
                    "version": "dev-master",
                    "require": { "a": "^3.0 || ^4.0@dev" },
                    "extra": { "branch-alias": { "dev-master": "4.x-dev" } }
                }
            ]
        }
    ],
    "require": {
        "a": "^4.0@dev",
        "b": "^4.0@dev"
    }
}

--RUN--
install

--EXPECT--
Installing a (4.x-dev)
Installing b (4.x-dev)

--EXPECT-EXIT-CODE--
0
```

```
There was 1 failure:

1) Composer\Test\InstallerTest::testIntegration with data set "github-issues-5505.test" ('github-issues-5505.test', '\nSee Github issue #5505 ( git...505 ).', null, array(array(array('package', array(array('a', 'dev-master', array('^3.0 || ^4.0@dev'), array('<3.0 || >=5.0'), array(array('4.x-dev'))), array('b', 'dev-master', array('^3.0 || ^4.0@dev'), array(array('4.x-dev')))))), array('^4.0@dev', '^4.0@dev')), array(), array(), 'install', array(), null, 'Installing a (4.x-dev)\nInstal...x-dev)', 0)
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for a ^4.0@dev -> satisfiable by a[4.x-dev].
    - don't install b 4.x-dev|install b dev-master
    - b dev-master conflicts with a[4.x-dev].
    - Installation request for b ^4.0@dev -> satisfiable by b[4.x-dev].
```

**Comment by alcohol** [comment]
@naderman you know more about the solver than me. What do you think we can do about this?

**Comment by naderman** [comment]
Off the top of my head this doesn't have much to do with the solver but with how we interpret conflict statements regarding aliases, but will have to spend a bit more time to really understand this.

**Comment by Seldaek** [comment]
Yup the problem seems to be that for requires, we allow either dev-master OR the alias to work, but for conflict as well it conflicts with either dev-master OR the alias, while for conflict it really only shouldn't conflict if the alias allows it, but I'm not sure this is solvable. Aliases are a convenience but they are also a hack to some extent and making them work in all edge cases might not be feasible reasonably.

Maybe you can drop the `>=5` conflict, or turn it into a `>=5 && <99`, that should solve it I assume.

**Comment by core23** [comment]
Thank you for you analysis. Dropping the conflict rule was my first idea to solve this issue temporary. 

The question is, if this is a bug or `works as designed`. Are you going to fix this?

**Comment by Seldaek** [comment]
As I hinted at, I am not sure that we can solve it, at least within a reasonable amount of time. So I'd rather you use a workaround like `>=5 && <999` which is safe enough, if it works :)

**Comment by slbmeh** [comment]
In https://github.com/composer/composer/blob/master/src/Composer/DependencyResolver/Pool.php#L328 it uses `Package::getVersion()`, and that gives the `9999999-dev` from `composer/semver` version parser.  This in particular may be solver related, but I noticed the same problem in #5328 which is why I did the commit to show the pretty version output in the status command.  Perhaps a service that can determine the eligible versions for a package may be in order.  Or at the very least the pool needs to be aware of alternative/equivalent versions for constraint comparison.

**Comment by greg0ire** [comment]
Looks like `&&` is not recognized, the docs says to use `,` or ` ` (a space) if anyone is wondering.

**Comment by greg0ire** [comment]
No problem, there are good docs :)

**Comment by Soullivaneuh** [comment]
I have a similar problem, I think it's related to this issue. Could you please confirm?

With this simple `composer.json` file:

```json
{
    "require": {
        "php": "^5.6 || ^7.0"
    },
    "conflict": {
        "silex/silex": "<1.2 || >=3.0"
    }
}
```

If I run this command:

```bash
composer require "silex/silex:dev-master" # 2.1.x-dev ATM
```

I got:

```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for __root__ No version set (parsed as 1.0.0) -> satisfiable by __root__[No version set (parsed as 1.0.0)].
    - silex/silex dev-master conflicts with __root__[No version set (parsed as 1.0.0)].
    - Installation request for silex/silex dev-master -> satisfiable by silex/silex[dev-master].


Installation failed, reverting ./composer.json to its original content.
```

@Seldaek the `<999` is not reliable IMHO because if I try this:

```json
{
    "require": {
        "php": "^5.6 || ^7.0"
    },
    "conflict": {
        "silex/silex": "<1.2 || >=2.0 <999"
    }
}
```

I should not be able to install any 2.x version of Silex, because of `>=2.0` instead of `>=3.0`.

This works with `2.1.x-dev`

```
$ composer require "silex/silex:2.1.x-dev"
    1/2:	http://packagist.org/p/provider-2017-01$1aeaf0a8f157d236ec6edc3efd14d11dee47446b641c1d8ce8646aae9bcf20a1.json
    2/2:	http://packagist.org/p/provider-latest$73a89701006ae7ed2cf85b768fd0bb6963f38df02304709572351e88052766e2.json
    Finished: success: 2, skipped: 0, failure: 0, total: 2
./composer.json has been updated
    1/2:	http://packagist.org/p/provider-2017-01$1aeaf0a8f157d236ec6edc3efd14d11dee47446b641c1d8ce8646aae9bcf20a1.json
    2/2:	http://packagist.org/p/provider-latest$73a89701006ae7ed2cf85b768fd0bb6963f38df02304709572351e88052766e2.json
    Finished: success: 2, skipped: 0, failure: 0, total: 2
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - silex/silex 2.1.x-dev conflicts with __root__[No version set (parsed as 1.0.0)].
    - __root__ No version set (parsed as 1.0.0) conflicts with silex/silex[2.1.x-dev].
    - silex/silex 2.1.x-dev conflicts with __root__[No version set (parsed as 1.0.0)].
    - Installation request for __root__ No version set (parsed as 1.0.0) -> satisfiable by __root__[No version set (parsed as 1.0.0)].
    - Installation request for silex/silex 2.1.x-dev -> satisfiable by silex/silex[2.1.x-dev].
```

But not with `dev-master`:

```
$ composer require "silex/silex:dev-master"
    1/1:	http://packagist.org/p/provider-latest$f0d8fa263e7ffbce16f6ef534b4e13acf509898704a6da24522d7b29c64a8ce5.json
    Finished: success: 1, skipped: 0, failure: 0, total: 1
./composer.json has been updated
    1/1:	http://packagist.org/p/provider-latest$f0d8fa263e7ffbce16f6ef534b4e13acf509898704a6da24522d7b29c64a8ce5.json
    Finished: success: 1, skipped: 0, failure: 0, total: 1
Loading composer repositories with package information
Updating dependencies (including require-dev)
Generating autoload files
No changelogs summary
1 package is not up to date:

  - silex/silex (2.1.x-dev) latest is dev-master
```

The package is installed, but it should not be.

**Comment by naderman** [comment]
@Soullivaneuh the purpose of the require command is to overwrite the require constraint for the named package in your composer.json, so what you have in your composer.json for the package you are requiring doesn't matter, it'll simply get overwritten.

**Comment by Soullivaneuh** [comment]
> so what you have in your composer.json for the package you are requiring doesn't matter, it'll simply get overwritten.

If I understand your logic, then requiring `2.1.x-dev` should be accepted too, but it is not.

**Comment by alcohol** [comment]
Because `2.1.x-dev` conflicts with your conflict rules..?

**Comment by Seldaek** [comment]
This is fixed in Composer 2.0
